src/main/kotlin/org/codefreak/codefreak/entity/Grade.kt

```diff
@@ -0,0 +1,45 @@
+package org.codefreak.codefreak.entity
+
+import javax.persistence.CascadeType
+import javax.persistence.Entity
+import javax.persistence.FetchType
+import javax.persistence.JoinColumn
+import javax.persistence.ManyToOne
+import javax.persistence.OneToMany
+import javax.persistence.OneToOne
+
+@Entity
+class Grade(
+
+  /**
+   * Every Evaluation can have a grade. Depends if enabled or not. Later we can pick the a result
+   */
+  @OneToOne
+  @JoinColumn(name = "evaluation", referencedColumnName = "id")
+  var evaluation: Evaluation,
+
+  /**
+   * A Grade is always part of a set in answer entity.
+   */
+  @ManyToOne
+  var answer: Answer
```

Comment:
Every answer has a Grade. Otherwise every Evaluation had multiply grades and answers but which grade belongs to what answer?

Reply:
Every evalution belongs to a single answer && every grade belongs to a single evaluation => every grade belongs to a single answer. I don't see why we need an additional relation here 🤔

https://github.com/codefreak/codefreak/blob/f411cc31eecbb789ab071757c0c2135ed903ae8e/src/main/kotlin/org/codefreak/codefreak/entity/Evaluation.kt#L19

Reply:
Oh sorry, i mixed something up. Every Evaluation which has an Autograder turned on receives a grade. So if a student runs several evaluations on an answer there will be multiply evaluations and grades. We want to pick the best grade so i made a relation. Yes, i can access the grades through their evaluations. I thought it just would be simpler to access like i did with the GradeDefinition.
I will check the ER Model again and encapsulate everything properly. Didnt think it would be an issue.

In general i encountered a SQL-Deadlock in a previous project due to too much data and cascading. Kept several relations on fetchtype.LAZY and reduced the nessecity of especially long cascades i had to load everything manually, even unwanted entities, afterwards. Thats why i developed this foreign key behaviour. :p

Reply:
Ah, I see. Finding the best grade can be done on the database as we do with the _latest evaluation_
https://github.com/codefreak/codefreak/blob/7bd41e09b2eb005180f4e9a4f6fc4f43ce77553e/src/main/kotlin/org/codefreak/codefreak/repository/EvaluationRepository.kt#L11

In general I think this is primarily an issue of _premature optimization_. If there are indeed performance issues, I think they should be solved in another way. The same is true for lazy fetching of JPA entities. This is not always a good behavior and to turn it of we have to use custom JPQL queries.

I think you can leave it as is for now. @HenningCash and I will have a closer look on the ER model in general.

Reply:
Thats what i actually did in the GradeRepository.
` fun findFirstByAnswerIdOrderByGradePercentageDesc(id: UUID): Optional<Grade>`

Yes, i read the article you've posted. Insightful. :)
I'll keep the ER-Model as it is for now.